src/main/java/org/openstreetmap/atlas/geography/atlas/builder/RelationBean.java

```diff
@@ -333,50 +328,85 @@ public int hashCode()
     @Override
     public boolean isEmpty()
     {
-        return this.memberIdentifiers.isEmpty();
+        return this.beanItems.isEmpty();
     }
 
     @Override
     public Iterator<RelationBeanItem> iterator()
     {
-        final List<RelationBeanItem> result = new ArrayList<>();
-        for (int index = 0; index < this.memberIdentifiers.size(); index++)
-        {
-            result.add(new RelationBeanItem(this.memberIdentifiers.get(index),
-                    this.memberRoles.get(index), this.memberTypes.get(index)));
-        }
-        return result.iterator();
+        return this.beanItems.iterator();
     }
 
     public RelationBean merge(final RelationBean other)
     {
         return RelationBean.mergeBeans(this, other);
     }
 
+    /**
+     * Remove an item from this {@link RelationBean} with a given identifier and {@link ItemType}.
+     * If an item is actually removed, this method will return its role wrapped in an
+     * {@link Optional}. If nothing was removed, then the {@link Optional} will be empty.
+     * 
+     * @param identifier
+     *            the id to remove
+     * @param itemType
+     *            the type to remove
+     * @return an {@link Optional} wrapping the role of the removed item
+     */
+    public Optional<String> removeItem(final Long identifier, final ItemType itemType)
```

Comment:
Shouldn't it remove all the roles matching the same item? I know it is uncommon, but sometimes it happens to have the same object with different roles

Reply:
Good point. My initial thought was to leave it up to the user to deal with this. If they want to remove a specific role, then they should be using the `removeItem` that allows role specification. I can change the behavior here to do what you said, and then change the name to `removeAllMatchingItems` for clarity.